Remotion/Data/DomainObjects/Infrastructure/DomainObjectTransactionContextImplementation.cs

```diff
@@ -0,0 +1,103 @@
+    {
+      ArgumentUtility.CheckNotNull(nameof(domainObject), domainObject);
+      DomainObject = domainObject;
+    }
+
+    public bool IsReferenceInitializeEventExecution { get; set; }
+
+    public DomainObject DomainObject { get; }
+
+    public object? GetTimestamp (ClientTransaction clientTransaction)
+    {
+      ArgumentUtility.CheckNotNull(nameof(clientTransaction), clientTransaction);
+      DomainObjectCheckUtility.CheckIfRightTransaction(DomainObject, clientTransaction);
+
+      ThrowIfReferenceInitializeEventExecuting();
+      return clientTransaction.DataManager.GetDataContainerWithLazyLoad(DomainObject.ID, true)!.Timestamp;
+    }
+
+    public DomainObjectState GetState (ClientTransaction clientTransaction)
+    {
+      ArgumentUtility.CheckNotNull(nameof(clientTransaction), clientTransaction);
+      DomainObjectCheckUtility.CheckIfRightTransaction(DomainObject, clientTransaction);
+
+      ThrowIfReferenceInitializeEventExecuting();
+      return clientTransaction.DataManager.GetState(DomainObject.ID);
+    }
+
+    public void RegisterForCommit (ClientTransaction clientTransaction)
+    {
+      ArgumentUtility.CheckNotNull(nameof(clientTransaction), clientTransaction);
+      DomainObjectCheckUtility.CheckIfRightTransaction(DomainObject, clientTransaction);
+
+      ThrowIfReferenceInitializeEventExecuting();
+      var dataContainer = clientTransaction.DataManager.GetDataContainerWithLazyLoad(DomainObject.ID, true)!;
+      if (dataContainer.State.IsDeleted)
+        return;
+
+      if (dataContainer.State.IsNew)
+        return;
+
+      dataContainer.MarkAsChanged();
+    }
+
+    public void EnsureDataAvailable (ClientTransaction clientTransaction)
+    {
+      ArgumentUtility.CheckNotNull(nameof(clientTransaction), clientTransaction);
+      DomainObjectCheckUtility.CheckIfRightTransaction(DomainObject, clientTransaction);
+
+      ThrowIfReferenceInitializeEventExecuting();
+      clientTransaction.EnsureDataAvailable(DomainObject.ID);
+
+      DataContainer? dataContainer;
+      Assertion.DebugAssert(
+          (dataContainer = clientTransaction.DataManager.DataContainers[DomainObject.ID]) != null
+          && dataContainer.DomainObject == DomainObject,
+          "Guaranteed because CheckIfRightTransaction ensures that DomainObject is enlisted.");
+    }
+
+    public bool TryEnsureDataAvailable (ClientTransaction clientTransaction)
+    {
+      ArgumentUtility.CheckNotNull(nameof(clientTransaction), clientTransaction);
+      DomainObjectCheckUtility.CheckIfRightTransaction(DomainObject, clientTransaction);
+
+      return clientTransaction.TryEnsureDataAvailable(DomainObject.ID);
+    }
+
+    public void ThrowIfReferenceInitializeEventExecuting ()
+    {
+      if (IsReferenceInitializeEventExecution)
+        throw new InvalidOperationException("While the OnReferenceInitializing event is executing, this member cannot be used.");
```

Comment:
```suggestion
        throw new InvalidOperationException("While the DomainObject.OnReferenceInitializing event is executing, this member cannot be used.");
```

Reply:
@a-ctor we should play with branch prediction and check if placing the exception into the else-branch causes better performance. it's not clear if we can measure this, we may need to check the assembler. we want a version where no jump is needed in the event that no exception is thrown because that's the standard case. It may be that .net is already optimizing for this.